### ISSUE TITLE
Object destructuring broken: variables are bound to the wrong object

### ISSUE THREAD

**Issue by shovon** [Microsoft/TypeScript#9363]
**TypeScript Version:** 1.8.10

**Code**

```ts
interface ISomething {
	foo: string,
	bar: string
}

function foo({}, {foo, bar}: ISomething) {
	
}
```

**Expected behavior:**

```ts
function foo(_a, _b) {
    var foo = _b.foo, bar = _b.bar;
}
```

**Actual behavior:** 

```ts
function foo(_a, _b) {
    var foo = _a.foo, bar = _a.bar;
}
```

**Comment by mhegazy** [comment]
@yuit can you take a look.

**Comment by shovon** [comment]
@yuit fyi, in a function with two parameters, each representing objects with properties that can be destructured, it doesn't matter whether the client code expects something or nothing from any of these objects, if either null or undefined is passed in to either of these parameters, an exception should be raised.

Example:

```ts
function something({}, {foo, bar}) {
  // ...
}

function another(viaIdentifier, { foo, bar }) {
  // ...
}

something(undefined, { foo: 'foo', bar: 'bar' });
// This is an ERROR, because the first parameter is set to undefined, and the
// function `something` was declared to have the first parameter be
// destructured, even if we aren't asking for any properties from it

another(undefined, { foo: 'foo', bar: 'bar' });
// This is perfectly fine, because the first parameter is not destructured in
// function declaration of `another`
```

**Comment by yuit** [comment]
@shovon I got the fix for the first comment. I am investigating your second comment

**Comment by yuit** [comment]
@shovon I have another PR to fix your second comment, if you would like to take a look !
